File: Python/reporter.py
"""Console reporter for rule matching results."""

from models import RuleResult, TestFacts, MockClause


def print_test_summary(tests: list[TestFacts]) -> None:
    print(f"\n📋 Found {len(tests)} test functions\n")
    for t in tests:
        type_counts: dict[str, int] = {}
        for f in t.facts:
            name = type(f).__name__.replace("Fact", "")
            type_counts[name] = type_counts.get(name, 0) + 1
        counts_str = ", ".join(f"{v} {k}" for k, v in type_counts.items())
        print(f"   • {t.test_name}")
        print(f"     Facts: {counts_str}")


def print_results(results: list[RuleResult]) -> None:
    print()
    print("═══════════════════════════════════════════════════════════")
    print("📊 RULE VERIFICATION (Constraint Satisfaction)")
    print("═══════════════════════════════════════════════════════════\n")

    # Build parent→children lookup
    children_of: dict[str, list[RuleResult]] = {}
    child_ids: set[str] = set()
    for r in results:
        if r.rule.parent:
            children_of.setdefault(r.rule.parent, []).append(r)
            child_ids.add(r.rule.id)

    for r in results:
        if r.rule.id in child_ids:
            continue
        _print_single_result(r, indent="")
        if r.rule.id in children_of:
            children = children_of[r.rule.id]
            print(f"   📎 Edge cases ({len(children)}):")
            for child in children:
                _print_single_result(child, indent="   ")

    implemented = sum(1 for r in results if r.is_implemented)
    print("═══════════════════════════════════════════════════════════")
    print(f"Summary: {implemented}/{len(results)} rules implemented")

    _print_category_summary(results, "Raises",
        lambda r: any(a.semantic_meaning.lower() == "raises" for a in r.rule.asserts))
    _print_category_summary(results, "Mocks",
        lambda r: any(isinstance(g, MockClause) for g in r.rule.given))

    print()
    print("── Implementation Counts ───────────────────────────────")
    for r in results:
        prefix = "  └─ " if r.rule.parent else ""
        count = len(r.all_matched_tests)
        bar = "█" * count + ("░" if count == 0 else "")
        print(f"   {prefix}{r.rule.id:<30} {bar} {count}")
    print("═══════════════════════════════════════════════════════════")


def _print_single_result(r: RuleResult, indent: str) -> None:
    parent_label = f" (edge case of {r.rule.parent})" if r.rule.parent else ""
    if r.is_implemented:
        print(f"{indent}✅ IMPLEMENTED: {r.rule.id}{parent_label}")
        print(f"{indent}   {r.rule.description}")
        print(f"{indent}   Test: {r.matched_test}")
        if r.bindings:
            bind_str = ", ".join(f"{k}={v}" for k, v in r.bindings.items())
            print(f"{indent}   Bindings: {bind_str}")
        print(f"{indent}   ✓ {len(r.given_matched)} given + {len(r.assert_matched)} assert matched")
        print(f"{indent}   🔢 Implemented {len(r.all_matched_tests)} time(s) across tests:")
        for t in r.all_matched_tests:
            print(f"{indent}      • {t}")
    else:
        print(f"{indent}❌ NOT IMPLEMENTED: {r.rule.id}{parent_label}")
        print(f"{indent}   {r.rule.description}")
        if r.given_missing:
            print(f"{indent}   Missing given:")
            for g in r.given_missing:
                print(f"{indent}      • {g}")
        if r.assert_missing:
            print(f"{indent}   Missing assert:")
            for a in r.assert_missing:
                print(f"{indent}      • {a}")
    print()


def _print_category_summary(results: list[RuleResult], label: str, predicate) -> None:
    category = [r for r in results if predicate(r)]
    if not category:
        print(f"   {label}: N/A (no rules)")
        return
    impl = sum(1 for r in category if r.is_implemented)
    pct = round(100.0 * impl / len(category))
    print(f"   {label}: {impl}/{len(category)} implemented ({pct}%)")


def print_facts(tests: list[TestFacts]) -> None:
    from models import CreatedFact, CalledFact, AssignedFact, MockSetupFact, AssertedFact

    print("\n── Extracted facts (debug) ──────────────────────────────\n")
    for t in tests:
        print(f"Test: {t.test_name}")
        for f in t.facts:
            if isinstance(f, CreatedFact):
                desc = f"  [{f.sequence}] Created {f.variable}: {f.type_name}({', '.join(f.arguments)})"
            elif isinstance(f, CalledFact):
                recv = f.receiver or "(static)"
                desc = f"  [{f.sequence}] Called {recv}.{f.method}({', '.join(f.arguments)})"
                if f.result_variable:
                    desc += f" → {f.result_variable}"
                if f.receiver_type:
                    desc += f" [type: {f.receiver_type}]"
                if f.callback_assign_target:
                    desc += f" [cb→{f.callback_assign_target}]"
                if f.callback_throw_type:
                    desc += f" [throws:{f.callback_throw_type}]"
            elif isinstance(f, AssignedFact):
                desc = f"  [{f.sequence}] Assigned {f.target} = {f.source}"
                if f.cast_type:
                    desc += f" as {f.cast_type}"
            elif isinstance(f, MockSetupFact):
                ret = f.return_value if f.return_value else f"side_effect={f.side_effect}"
                desc = f"  [{f.sequence}] Mock {f.mock_variable}.{f.method} returns {ret}"
            elif isinstance(f, AssertedFact):
                desc = f"  [{f.sequence}] Assert {f.semantic_meaning}({', '.join(f.arguments)})"
            else:
                desc = f"  [{f.sequence}] {f}"
            print(desc)
        if t.variable_types:
            types_str = ", ".join(f"{k}:{v}" for k, v in t.variable_types.items())
            print(f"  Variable types: {types_str}")
        print()

File: CSharp/Services/Reporter.cs
/// <summary>Console reporter for rule matching results.</summary>
static class Reporter
{
    public static void PrintTestSummary(List<TestFacts> tests)
    {
        Console.WriteLine($"\n📋 Found {tests.Count} test methods\n");
        foreach (var t in tests)
        {
            var factCounts = string.Join(", ",
                t.Facts.GroupBy(f => f.GetType().Name.Replace("Fact", ""))
                       .Select(g => $"{g.Count()} {g.Key}"));
            Console.WriteLine($"   • {t.TestName}");
            Console.WriteLine($"     Facts: {factCounts}");
        }
    }

    public static void PrintResults(List<RuleResult> results)
    {
        Console.WriteLine();
        Console.WriteLine("═══════════════════════════════════════════════════════════");
        Console.WriteLine("📊 RULE VERIFICATION (Constraint Satisfaction)");
        Console.WriteLine("═══════════════════════════════════════════════════════════\n");

        // Build parent→children lookup
        var childrenOf = results
            .Where(r => r.Rule.Parent != null)
            .GroupBy(r => r.Rule.Parent!)
            .ToDictionary(g => g.Key, g => g.ToList());
        var childIds = new HashSet<string>(results.Where(r => r.Rule.Parent != null).Select(r => r.Rule.Id));

        // Build OR group lookup
        var orGroups = results
            .Where(r => r.Rule.OrGroup != null)
            .GroupBy(r => r.Rule.OrGroup!)
            .ToDictionary(g => g.Key, g => g.ToList());
        var orGroupIds = new HashSet<string>(results.Where(r => r.Rule.OrGroup != null).Select(r => r.Rule.Id));
        var printedOrGroups = new HashSet<string>();

        foreach (var r in results)
        {
            // Skip children here; they are printed under their parent
            if (childIds.Contains(r.Rule.Id)) continue;

            // Handle OR groups: print as a group when first member is encountered
            if (r.Rule.OrGroup != null)
            {
                if (printedOrGroups.Contains(r.Rule.OrGroup)) continue;
                printedOrGroups.Add(r.Rule.OrGroup);
                PrintOrGroup(r.Rule.OrGroup, orGroups[r.Rule.OrGroup]);
                continue;
            }

            PrintSingleResult(r, indent: "");

            // Print children indented
            if (childrenOf.TryGetValue(r.Rule.Id, out var children))
            {
                Console.WriteLine($"   📎 Edge cases ({children.Count}):");
                foreach (var child in children)
                    PrintSingleResult(child, indent: "   ");
            }
        }

        // Summary: OR groups count as 1
        var orGroupCount = orGroups.Count;
        var orGroupImplemented = orGroups.Values.Count(g => g.Any(r => r.IsImplemented));
        var standalone = results.Where(r => r.Rule.OrGroup == null).ToList();
        var totalLogical = standalone.Count + orGroupCount;
        var totalImplemented = standalone.Count(r => r.IsImplemented) + orGroupImplemented;

        Console.WriteLine("═══════════════════════════════════════════════════════════");
        Console.WriteLine($"Summary: {totalImplemented}/{totalLogical} rules implemented");
        if (orGroupCount > 0)
            Console.WriteLine($"   ({orGroupCount} OR group(s), each counting as 1 rule)");

        // Category breakdowns
        PrintCategorySummary(results, "Throws",
            r => r.Rule.Assert.Any(a => a.SemanticMeaning.Equals("Throws", StringComparison.OrdinalIgnoreCase)));
        PrintCategorySummary(results, "Mocks",
            r => r.Rule.Given.Any(g => g is MockClause));

        Console.WriteLine();
        Console.WriteLine("── Implementation Counts ───────────────────────────────");
        var printedOrGroupsBars = new HashSet<string>();
        foreach (var r in results)
        {
            // OR group members: show group line once, then individual alternatives indented
            if (r.Rule.OrGroup != null)
            {
                if (printedOrGroupsBars.Contains(r.Rule.OrGroup)) continue;
                printedOrGroupsBars.Add(r.Rule.OrGroup);
                var groupMembers = orGroups[r.Rule.OrGroup];
                var groupCount = groupMembers.Sum(m => m.AllMatchedTests.Count);
                var groupBar = new string('█', groupCount) + (groupCount == 0 ? "░" : "");
                Console.WriteLine($"   {"[OR] " + r.Rule.OrGroup,-30} {groupBar} {groupCount}");
                foreach (var m in groupMembers)
                {
                    var mCount = m.AllMatchedTests.Count;
                    var mBar = new string('█', mCount) + (mCount == 0 ? "░" : "");
                    Console.WriteLine($"     └─ {m.Rule.Id,-28} {mBar} {mCount}");
                }
                continue;
            }

            var prefix = r.Rule.Parent != null ? "  └─ " : "";
            var count = r.AllMatchedTests.Count;
            var bar = new string('█', count) + (count == 0 ? "░" : "");
            Console.WriteLine($"   {prefix}{r.Rule.Id,-30} {bar} {count}");
        }
        Console.WriteLine("═══════════════════════════════════════════════════════════");
    }

    static void PrintOrGroup(string groupName, List<RuleResult> alternatives)
    {
        var anyImplemented = alternatives.Any(r => r.IsImplemented);
        var icon = anyImplemented ? "✅" : "❌";
        var status = anyImplemented ? "IMPLEMENTED" : "NOT IMPLEMENTED";

        Console.WriteLine($"{icon} OR GROUP: {groupName} — {status}");
        Console.WriteLine($"   Any one of {alternatives.Count} alternatives suffices:");

        foreach (var r in alternatives)
        {
            var altIcon = r.IsImplemented ? "✓" : "✗";
            Console.WriteLine($"   {altIcon} {r.Rule.Id}: {r.Rule.Description}");
            if (r.IsImplemented)
            {
                Console.WriteLine($"      Test: {r.MatchedTest}");
                if (r.Bindings != null)
                {
                    var bindStr = string.Join(", ", r.Bindings.Select(kv => $"{kv.Key}={kv.Value}"));
                    Console.WriteLine($"      Bindings: {bindStr}");
                }
                Console.WriteLine($"      🔢 Implemented {r.AllMatchedTests.Count} time(s)");
            }
        }
        Console.WriteLine();
    }

    static void PrintSingleResult(RuleResult r, string indent)
    {
        var parentLabel = r.Rule.Parent != null ? $" (edge case of {r.Rule.Parent})" : "";
        if (r.IsImplemented)
        {
            Console.WriteLine($"{indent}✅ IMPLEMENTED: {r.Rule.Id}{parentLabel}");
            Console.WriteLine($"{indent}   {r.Rule.Description}");
            Console.WriteLine($"{indent}   Test: {r.MatchedTest}");
            if (r.Bindings != null)
            {
                var bindStr = string.Join(", ", r.Bindings.Select(kv => $"{kv.Key}={kv.Value}"));
                Console.WriteLine($"{indent}   Bindings: {bindStr}");
            }
            Console.WriteLine($"{indent}   ✓ {r.GivenMatched.Count} given + {r.AssertMatched.Count} assert matched");
            Console.WriteLine($"{indent}   🔢 Implemented {r.AllMatchedTests.Count} time(s) across tests:");
            foreach (var t in r.AllMatchedTests)
                Console.WriteLine($"{indent}      • {t}");
        }
        else
        {
            Console.WriteLine($"{indent}❌ NOT IMPLEMENTED: {r.Rule.Id}{parentLabel}");
            Console.WriteLine($"{indent}   {r.Rule.Description}");
            if (r.GivenMissing.Count > 0)
            {
                Console.WriteLine($"{indent}   Missing given:");
                foreach (var g in r.GivenMissing) Console.WriteLine($"{indent}      • {g}");
            }
            if (r.AssertMissing.Count > 0)
            {
                Console.WriteLine($"{indent}   Missing assert:");
                foreach (var a in r.AssertMissing) Console.WriteLine($"{indent}      • {a}");
            }
        }
        Console.WriteLine();
    }

    static void PrintCategorySummary(List<RuleResult> results, string label, Func<RuleResult, bool> predicate)
    {
        var category = results.Where(predicate).ToList();
        if (category.Count == 0)
        {
            Console.WriteLine($"   {label}: N/A (no rules)");
            return;
        }
        var impl = category.Count(r => r.IsImplemented);
        var pct = (int)Math.Round(100.0 * impl / category.Count);
        Console.WriteLine($"   {label}: {impl}/{category.Count} implemented ({pct}%)");
    }

    public static void PrintFacts(List<TestFacts> tests)
    {
        Console.WriteLine("\n── Extracted facts (debug) ──────────────────────────────\n");
        foreach (var t in tests)
        {
            Console.WriteLine($"Test: {t.TestName}");
            foreach (var f in t.Facts)
            {
                var desc = f switch
                {
                    CreatedFact c => $"  [{c.Sequence}] Created {c.Variable}: {c.Type}({string.Join(", ", c.Arguments)})",
                    CalledFact c => $"  [{c.Sequence}] Called {c.Receiver ?? "(static)"}.{c.Method}({string.Join(", ", c.Arguments)})" +
                                   (c.ResultVariable != null ? $" → {c.ResultVariable}" : "") +
                                   (c.ReceiverType != null ? $" [type: {c.ReceiverType}]" : "") +
                                   (c.CallbackAssignTarget != null ? $" [cb→{c.CallbackAssignTarget}]" : "") +
                                   (c.CallbackThrowType != null ? $" [throws:{c.CallbackThrowType}]" : ""),
                    AssignedFact a => $"  [{a.Sequence}] Assigned {a.Target} = {a.Source}" + (a.CastType != null ? $" as {a.CastType}" : ""),
                    MockSetupFact m => $"  [{m.Sequence}] Mock {m.MockVariable}.{m.Method} returns {m.ReturnValue ?? $"throws {m.ThrowsType}"}",
                    AssertedFact a => $"  [{a.Sequence}] Assert {a.SemanticMeaning}({string.Join(", ", a.Arguments)})",
                    _ => f.ToString() ?? ""
                };
                Console.WriteLine(desc);
            }
            if (t.VariableTypes.Count > 0)
            {
                Console.WriteLine($"  Variable types: {string.Join(", ", t.VariableTypes.Select(kv => $"{kv.Key}:{kv.Value}"))}");
            }
            Console.WriteLine();
        }
    }
}
